### ISSUE TITLE
unreachable false positive in pyright 1.1.370

### ISSUE THREAD

**Issue by DetachHead** [DetachHead/basedpyright#467]
```py
from typing import final


@final
class A: ...


@final
class B: ...

def func2(x: type[A] | type[B] | None | int, y: type[A]):
    if x is not y:
        reveal_type(x, expected_text="type[B] | int | None")
    else:
        reveal_type(x, expected_text="type[A]") # unreachable
```
